src/lex/cpp.rs

```diff
@@ -1152,6 +1158,25 @@ impl<'a> PreProcessor<'a> {
         }
         Ok(ret)
     }
+
+    fn update_builtin_definitions(&mut self) {
+        use Definition::Object;
+        let int_def = |i: i32| Object(vec![LiteralToken::Int(RcStr::from(i.to_string())).into()]);
+        let str_def = |s: &str| {
+            let rcstr = RcStr::from(format!("\"{}\"", s));
+            Object(vec![LiteralToken::Str(vec![rcstr]).into()])
+        };
+        self.definitions.extend(map! {
+            "__LINE__".into() => int_def((self.line() + 1) as i32),
+            "__FILE__".into() => str_def(self.file_processor.path().to_str().unwrap()),  // Fails if non-Unicode
```

Comment:
Non Unicode file paths are recoverable I think. This defines it to be empty if it isn't valid UTF-8. Another alternative is to use `to_str_lossy`, but that might not be supported by RcStr.
```suggestion
            "__FILE__".into() => str_def(self.file_processor.path().to_str().unwrap_or_default()),  // Fails if non-Unicode
```

Reply:
Anything other than panicking would give the wrong behaviour, no? I can do `_or_default()` if you'd like though (I don't think `RcStr` will take lossy).

Reply:
Hmm, can you make it an error instead of a panic then? I don't think we should panic on user errors except in extreme cases (e.g. https://github.com/jyn514/saltwater/blob/master/src/ir/mod.rs#L73). See https://github.com/jyn514/saltwater/blob/0e36d97ac31a1875d6a1276b1f8840fedcef65e8/src/lex/cpp.rs#L442 for an example.

Reply:
I can do `_lossy` but I'm not sure how that works since I'm `format!`ing it into quotation marks before it goes to `RcStr` so I'm not sure if this does what you want. Also, can paths have quotation marks in them?